### PR TITLE
HDDS-8061. [Snapshot] Add SnapDiff client option to force a full diff

### PR DESCRIPTION
**Pull request by targetoee**
## What changes were proposed in this pull request?
By default, snapdiff would use the optimized DAG-based pruning approach. This change provides a full diff option of snapshot. 

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8061

## How was this patch tested?
unit test

### CODE REVIEW

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto [hunk 1]
```diff
@@ -1694,6 +1694,7 @@ message SnapshotDiffRequest {
   optional string toSnapshot = 4;
   optional string token = 5;
   optional uint32 pageSize = 6;
+  optional bool fullDiff = 7;
```
Comment (smengcl) [inline hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:1697]:
```suggestion
  optional bool forceFullDiff = 7;
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/SnapshotDiffHandler.java [hunk 2]
```diff
@@ -54,6 +54,11 @@ public class SnapshotDiffHandler extends Handler {
           " (optional)")
   private int pageSize;
 
+  @CommandLine.Option(names = {"-fd", "--use-full-diff"},
```
Comment (smengcl) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/SnapshotDiffHandler.java:57]:
```suggestion
  @CommandLine.Option(names = {"--ffd", "--force-full-diff"},
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/SnapshotDiffHandler.java [hunk 3]
```diff
@@ -54,6 +54,11 @@ public class SnapshotDiffHandler extends Handler {
...
+      description = "perform full diff of snapshot without using" +
+          " optimised DAG based pruning approach (optional)")
+  private boolean fullDiff;
```
Comment (smengcl) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/SnapshotDiffHandler.java:60]: This option would be useful for debugging. But for now let's hide this option by default to regular users (still usable). We can decide whether to unhide this later.

```suggestion
      description = "perform full diff of snapshot without using" +
          " optimised DAG based pruning approach (optional)",
      hidden = true)
  private boolean fullDiff;
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 4]
```diff
@@ -389,11 +393,13 @@ private Set<String> getDeltaFiles(OmSnapshot fromSnapshot,
       // End of Workaround
     }
 
-    if (deltaFiles.isEmpty()) {
+    if (useFullDiff || deltaFiles.isEmpty()) {
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:396]: This new condition looks reasonable to me.

But I'd like @sadanand48 or @hemantk-12 to review this change in this class as well.

**Review by smengcl — COMMENTED** [review]
Thanks @targetoee for the patch. Overall lgtm.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java (hunk 4 above)
Comment (sadanand48) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:396]: If useFullDiff is true then deltaFiles.isEmpty() is always true but I guess its ok to have this .

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java [hunk 5]
```diff
@@ -1059,11 +1059,13 @@ List<OzoneSnapshot> listSnapshot(String volumeName, String bucketName)
    * @param toSnapshot The name of the ending snapshot
    * @param token to get the index to return diff report from.
    * @param pageSize maximum entries returned to the report.
+   * @param forceFullDiff show full diff without using the optimised DAG method
```
Comment (smengcl) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java:1062]:
```suggestion
   * @param forceFullDiff request to force full diff, skipping DAG optimization
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OzoneManagerProtocol.java [hunk 6]
```diff
@@ -705,6 +705,7 @@ default List<SnapshotInfo> listSnapshot(String volumeName, String bucketName)
    * @param toSnapshot The name of the ending snapshot
    * @param token to get the index to return diff report from.
    * @param pageSize maximum entries returned to the report.
+   * @param forceFullDiff perform full diff without using optimized DAG approach
```
Comment (smengcl) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OzoneManagerProtocol.java:708]:
```suggestion
   * @param forceFullDiff request to force full diff, skipping DAG optimization
```

**Review by smengcl — COMMENTED** [review]
Thanks @targetoee . Latest changes lgtm apart minor javadoc comments.

**Comment by targetoee** [comment]
Done. Thanks @smengcl @sadanand48 for the review!

**Review by smengcl — APPROVED** [review]

**Comment by smengcl** [comment]
Thanks @targetoee for the patch. Thanks @sadanand48 for reviewing this as well.

**Pull request merged**